File: c/difference-of-squares/difference_of_squares.c
#include "difference_of_squares.h"

#include <math.h>
#include <stdint.h>

#define u32 uint32_t

u32 sum_of_squares(const u32 number) {

    u32 sum = 0;

    for (u32 i = 1; i <= number; i++) {
        sum += (u32)pow(i, 2);
    }

    return sum;
}

u32 square_of_sum(const u32 number) {

    u32 sum = 0;

    for (u32 i = 1; i <= number; i++) {
        sum += i;
    }

    return (u32)pow(sum, 2);
}

u32 difference_of_squares(const u32 number) {
    return square_of_sum(number) - sum_of_squares(number);
}

File: zig/difference-of-squares/difference_of_squares.zig
// Not sure if I should be testing for overflow
// but all tests pass.

const std = @import("std");
const math = std.math;

pub fn squareOfSum(number: usize) usize {
    var sum: usize = 0;
    var i: usize = 1;

    while (i <= number) : (i += 1) {
        sum += i;
    }

    return math.pow(usize, sum, 2);
}

pub fn sumOfSquares(number: usize) usize {
    var sum: usize = 0;
    var i: usize = 1;

    while (i <= number) : (i += 1) {
        sum += math.pow(usize, i, 2);
    }

    return sum;
}

pub fn differenceOfSquares(number: usize) usize {
    return squareOfSum(number) - sumOfSquares(number);
}
